Editor: Support VideoPress post-editor preview

mjuhasz opened the pull request:
Closes: #1730

This pull request enables previewing of VideoPress videos in the post editor.

There is currently no VideoPress preview offered in the post editor on WordPress.com. When adding a VideoPress video to a post, only the shortcode text is shown in the editor post content. Only after saving and previewing the post can a user see how the post will look like.

##### Before:

![screen shot 2016-01-18 at 10 50 45 am](https://cloud.githubusercontent.com/assets/1017839/12388184/a6413922-bdd1-11e5-878e-4578f418e730.png)

##### After:

![screen shot 2016-01-18 at 10 52 47 am](https://cloud.githubusercontent.com/assets/1017839/12388188/ac08cd0c-bdd1-11e5-8299-7cab09a5eccb.png)

##### Testing instructions:

###### Scenario 1 - adding media from your library:

1. Navigate to the Calypso post editor.
2. Select a site, if prompted.
3. Create a new post.
4. Click on Add Media, select a video from your library, then click Insert.
5. Note that a video player is displayed in the editor.
6. Confirm that you can start playing the video and adjust the volume.
7. Confirm that by having the video player in focus (e.g. after clicking at its top bar) a remove button shows up with which the video can be removed.

Strictly speaking Scenario 1 resolves issue #1730, nevertheless the following use cases should also work:

###### Scenario 2 - adding media by pasting a shortcode:

Same as Scenario 1, except Step 4: Insert the following shortcode text into the editor: `[wpvideo kUJmAcSf]`

###### Scenario 3 - using shortcode attributes:

Same as Scenario 1, except:

Step 4: Insert the following shortcode text into the editor: `[wpvideo kUJmAcSf at=90 ]`
Step 5: Note that a video player is displayed in the editor and its start time is 1m30s.

Remark: not all shortcode attributes are fully supported at this time for preview.

mjuhasz commented:
This PR solves #1730 in a way different from PR https://github.com/Automattic/wp-calypso/pull/2160: by rendering the Videopress iframe directly.

The aforementioned PR has the advantage of rendering the shortcode itself for preview so
* it accepts the same shortcode attributes (w, h, etc.)
* it renders the preview using the same dimensions as what the user sees when the Preview button is pressed

Parsing out the `video_guid` is also quite dodgy, there must be a better way to get it.

mjuhasz commented:
A few outstanding issues/remarks:

* default attributes are not left out from the url - they do not do any harm but they could be filtered out if we have definite defaults
* video width/height and aspect ratio are given some default value but ideally they would be queried from VideoPress. In Jetpack's `VideoPress_Video` class there is [a remote call](https://github.com/Automattic/jetpack/blob/417b533ff107593c6dec7e637689b88ea434edcd/modules/videopress/class.videopress-video.php#L299) to get these attributes. Similar logic could be replicated in the `wpview` plugin.
* `width` shortcode attribute is ignored right now because a [tinymce css entry](https://github.com/Automattic/wp-calypso/blob/master/client/components/tinymce/iframe.scss#L108) overrides it: the VideoPress iframe is a descendant of a div with class `wpview-wrap` so it gets 100% width. We did not have this issue when the shortcode was rendered because there was an extra iframe around the VideoPress one.
* some attributes are not parsed: `permalink` - it does not seem to take effect; `freedom` and `flashonly` - the player embedded in the VideoPress iframe is used

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,79 @@
+		const namedAttrs = shortcode.attrs.named;
+		const defaultWidth = 640;
+		const defaultHeight = defaultWidth * 9 / 16;
+		return {
```
Using `Object.assign` (or similarly, [`_.assign`](https://lodash.com/docs#assign) or [`_.defaults`](https://lodash.com/docs#defaults)) can be a handy way to clearly identify the default values, while still allowing override if value is known:

```js
return Object.assign( {
	w: defaultWidth,
	h: defaultHeight,
	at: 0,
	defaultLangCode: false
}, {
	videopress_guid: shortcode.attrs.numeric[0],
	w: parseInt( namedAttrs.w, 10 ),
	h: parseInt( namedAttrs.h, 10 ),
	autoplay: namedAttrs.autoplay === 'true',
	hd: namedAttrs.hd === 'true',
	loop: namedAttrs.loop === 'true',
	at: parseInt( namedAttrs.at, 10 ),
	defaultLangCode: namedAttrs.defaultlangcode
} );
```

Might be a negligible performance difference, though the clarity benefits should outweigh this. What do you think?

mjuhasz replied:
Very handy, indeed! It makes the intent of the code much clearer.

aduth replied:
>Very handy, indeed! It makes the intent of the code much clearer.

Actually, this will not work as expected, as the presence of the key in the second argument will override the default value, regardless of whether the value is truthy.

For example, try entering this in your browser console (assuming your browser supports `Object.assign`):

```js
Object.assign( { a: 1 }, { a: undefined } );
```

mjuhasz replied:
In the end I used lodash's [_.defaults](https://lodash.com/docs#defaults). I also had to substitute `NaN` for `undefined` to get the expected behavior (i.e. used the default if one exists for unspecified attributes).

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,79 @@
+
+	getEmbedUrl( attrs ) {
+		const nonUrlAttributes= ['videopress_guid', 'w', 'h'];
+		const str = Object.keys( attrs ).filter( function ( key ) {
```
[`_.omit`](https://lodash.com/docs#omit) might be a handy utility here.

mjuhasz replied:
Agreed, fixed!

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,79 @@
+				}
+			)
+		}).map(function ( key ) {
+			return encodeURIComponent( key ) + '=' + encodeURIComponent( attrs[key] );
```
Looks like you've reimplemented [`querystring.stringify`](https://nodejs.org/api/querystring.html#querystring_querystring_stringify_obj_sep_eq_options). You might consider leveraging it directly instead, as we're already using it elsewhere in the project.

mjuhasz replied:
Even less code to maintain :yum:

aduth commented:
Result of running `eslint client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx `:

```text
/Users/andrew/Documents/Code/wp-calypso/client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx
  43:36  warning  There must be a space inside this paren               space-in-parens
  49:25  warning  Infix operators must be spaced                        space-infix-ops
  50:52  warning  Unexpected space before function parentheses          space-before-function-paren
  51:44  warning  Unexpected space before function parentheses          space-before-function-paren
  52:6   warning  Expected indentation of 4 tab characters but found 5  indent
  53:5   warning  Expected indentation of 3 tab characters but found 4  indent
  55:5   warning  There must be a space inside this paren               space-in-parens
  55:10  warning  There must be a space inside this paren               space-in-parens
  55:18  warning  Unexpected space before function parentheses          space-before-function-paren
  57:5   warning  There must be a space inside this paren               space-in-parens
  57:11  warning  There must be a space inside this paren               space-in-parens
  57:15  warning  There must be a space inside this paren               space-in-parens

✖ 12 problems (0 errors, 12 warnings)
```

You might consider installing a linting plugin for your editor for these issues to be more evident:

https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#eslint

aduth commented on client/components/tinymce/iframe.scss:
```diff
@@ -103,11 +103,14 @@ a img {
 }
 
+.wpview-wrap iframe {
+  display: block;
```
Mixed whitespace (this should be tabs, not spaces).

Additionally, you may consider leveraging nesting for the pseudo-class `:not` selector ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#csssass-coding-guidelines)), i.e.

```sass
.wpview-wrap iframe {
	display: block;

	&:not( .wpview-type-video ) {
		width: 100%;
	}
}
```

Also note spaces between parens ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#spacing)).

mjuhasz replied:
Fixed.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view-container.jsx:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	fetchData( siteId ) {
+		setTimeout( () => {
```
So, from what I can tell, the purpose for querying/fetching media is to determine the dimensions for the embedded video? Can we not just default this to a fixed width value, similar to [what we're doing for images](https://github.com/Automattic/wp-calypso/blob/fb83eac4d214d85f978225172a57080da3e0996d/client/components/tinymce/plugins/media/restrict-size/index.js#L13)? To be clear, I'm not totally familiar with proportion variations for VideoPress videos, or to the exact extent `videopress-iframe.js` is capable of managing these details for us.

aduth replied:
It may also be worth exploring whether we should consider adding support for fetching individual video files from the REST API, such that we don't need to manage querying nor request the entire set of media.

See: https://developer.wordpress.com/docs/api/1.1/get/videos/%24guid/

aduth replied:
If you consider adding improvements for fetching video details from the REST API, you may want to look over our [newly proposed recommendations](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#current-recommendations) with regards to managing data.

mjuhasz replied:
Defaulting to a fixed width value would be the easiest. I chose not to do so because shortcodes can have a `w` and `h` (width and height, respectively) attribute that would be ignored in the preview but would take effect in the post preview/published post, so in the end, the in-editor preview would show something different than what we would see in the post preview/published post. It's a trade-off between having a not-so-realistic preview vs. adding code complexity to make the preview resemble the reality closer.

mjuhasz replied:
I will look into adding support for fetching video files by `videopress_guid` for two reasons:

* to eliminate complexity, that is, pagination and filtering
* to avoid having to create a store factory in order to fix the bug whereby the store mime-type query interferes with other consumers of the store. This bug manifests itself in not seeing anything but video files from your library if you hit Add Media in the post editor.

aduth commented:
I wonder if we could reuse some of the embed view logic to simplify this, as the oEmbed endpoint markup already returns the video's known dimensions and includes the resizing script automatically.

Example: https://public-api.wordpress.com/oembed/1.0/?url=https%3A%2F%2Fvideopress.com%2Fv%2FkUJmAcSf%3Fat%3D90&for=wordpress

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,124 @@
+			defaultLangCode: namedAttrs.defaultlangcode
+		};
+
+		return omit( attrs, ( v, k ) => defaultAttrValues[k] === v );
```
Many of the variable names in these files are obscure acronyms, which we try to avoid, as they're much more difficult to understand at a glance.

mjuhasz replied:
I made the variable names more descriptive.

Attribute names are cryptic because those are the names expected in the query string and I want to avoid mapping them back and forth.

aduth commented:
>in the end, the in-editor preview would show something different than what we would see in the post preview/published post.

Not to justify its existence, but there are many examples of this already in the post editor. We never account for a theme's `$content_width`, so many images and embeds will be displayed at an arbitrary size appropriate for the editor.

https://github.com/Automattic/wp-calypso/blob/eceb0b9/client/post-editor/style.scss#L82
https://github.com/Automattic/wp-calypso/blob/eceb0b9/client/components/tinymce/plugins/media/restrict-size/index.js#L13

I don't think we need worry too much about perfect accuracy between the post editor and the front-end, or at least not until we can tackle the problem consistently.

>I will look into adding support for fetching video files by `videopress_guid` for two reasons:

If you continue down this path, be sure to include the new fetching behavior as part of the [Redux state tree](https://github.com/Automattic/wp-calypso/tree/master/client/state#readme), which currently does not contain any media state.

https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#current-recommendations

mjuhasz commented:
I've added support to fetch individual video files (their metadata) from the REST API, and refactored the code to use the redux state tree.

I explored the option of using a fixed width but in order to avoid showing the video distorted we still need to set the height, and for that, we need the aspect ratio, and for that, the video metadata. 

Two things worth noting here:

* defaulting to a 16:9 ratio and calculating the preview video height from editor width and default aspect ratio  could be a way to tackle this, but having the correct ratio is obviously better user experience
* There are various ways to describe aspect ratio. What I as a user care about is the Display Aspect Ratio (DAR). What I am doing in the code is calculating the Storage Aspect Ratio (SAR), that is, the ratio of pixel dimension. This equals to DAR as long as Pixel Aspect Ratio (PAR) is 1 (that is, pixels are square). This is not true for DVDs for instance (720:576 ≠ 4:3), but it is true for hd source (1280:720 = 1920:1080 = 16:9). The right thing would be to preserve the DAR - that, however, is not returned from the wpcom endpoint, and I believe it is beyond the scope of this Github issue.

aduth commented on client/components/data/query-video/README.md:
```diff
@@ -0,0 +1,45 @@
+
+export default connect( ( state, props ) => {
+	return {
+		video: getVideo(state, props.guid)
```
Minor: Spaces between parens ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#spacing))

mjuhasz replied:
Fixed. (lint did not catch it in the markdown file, nor did I).

aduth commented on client/components/data/query-video/index.jsx:
```diff
@@ -0,0 +1,56 @@
+
+export default connect(
+	( state, props ) => {
+		const { guid } = props;
```
It's not of much concern, but there's really not a big advantage to destructing here, as it's more verbose than pulling the property off the object directly when there's only a single prop.

mjuhasz replied:
Fair enough. Fixed. A missed simplification after refactoring.

aduth commented on client/components/tinymce/plugins/wpcom-view/plugin.js:
```diff
@@ -90,11 +91,13 @@ function wpview( editor ) {
+				React.createElement( ReduxProvider, { store: editor.getParam( 'redux_store' ) },
+					React.createElement( views.components[ type ], {
+						content: getText( view ),
+						siteId: sites.getSelectedSite() ? sites.getSelectedSite().ID : null,
```
I'd recommend now that we're using Redux state, we retrieve the selected site in descendent connected components via [`getSelectedSite` state selector](https://github.com/Automattic/wp-calypso/blob/f83d11bc6e3cff8eea4fed32d41461a1eb7f200e/client/state/ui/selectors.js#L1-L13).

aduth replied:
Ah, I guess if other views rely on the site ID, we should continue to pass the prop. If you find the need to access it in your view component though, I'd use the above selector.

mjuhasz replied:
I am not using the site ID, other views are. Thank you for letting me know about the `getSelectedSite` selector!

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,140 @@
+}
+
+WpVideoView.propTypes = {
+	siteId: PropTypes.number,
```
You're not using this, so you ought remove it from `propTypes`. Define only those props the component itself needs, not what may or may not be passed.

mjuhasz replied:
Good to know. I was under the impression that all incoming properties have to be defined. Fixed.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	onLoad() {
+		const doc = tinymce.activeEditor.iframeElement.contentDocument;
```
I really wish we weren't pull the active editor from the global `tinymce` instance, though understandably it's not otherwise made available to this component. Can you do some digging to see if there's a way we can more explicitly pass the editor instance or iframe to the component?

mjuhasz replied:
Sure, I will do some digging.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,140 @@
+		return encodeURIComponent( content );
+	}
+
+	constructor( props ) {
```
I believe it's not necessary to define the `constructor` if you're not performing custom logic, and that the `super` constructor will be called by default.

mjuhasz replied:
Good point. Fixed.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,140 @@
+	getEmbedUrl( shortcodeAttributes ) {
+		const defaultAttributeValues = { hd: false, at: 0, defaultLangCode: undefined };
+		const attributesWithNonDefaultValues = omit( shortcodeAttributes, ( value, key ) => defaultAttributeValues[key] === value );
+		const queryString = QueryString.stringify( pick( attributesWithNonDefaultValues, ['autoplay', 'hd', 'loop', 'at', 'defaultLangCode'] ) );
```
We should watch the line length in this file. It hurts readability, and should be split into separate lines or variables when possible. See [spacing guidelines](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#spacing):

>Lines should usually be no longer than 80 characters, and should not exceed 100 (counting tabs as 4 spaces). This is a “soft” rule, but long lines generally indicate unreadable or disorganized code.

mjuhasz replied:
Fixed.

aduth commented on client/lib/wp/sync-handler/whitelist-handler.js:
```diff
@@ -8,7 +8,8 @@ const debug = debugFactory( 'calypso:sync-handler:whitelist' );
 	/^\/me\/settings/,
-	/^\/sites\/[\w.]+\/posts/
+	/^\/sites\/[\w.]+\/posts/,
+	/^\/videos/
```
I personally don't understand the implication of adding this to the whitelist. Can you explain or link me to a reference?

mjuhasz replied:
[Sync-handler](https://github.com/Automattic/wp-calypso/tree/master/client/lib/wp/sync-handler) is used for calypso offline. It intercepts client requests and stores server responses in local storage. It is work in progress (e.g. cache invalidation is not done, etc.) and I do not think that we have to whitelist the `videos` endpoint at this time. How offline video preview works (if it can work at all) can be evaluated and this endpoint whitelisted later. I removed it from the whitelist for the time being. (Earlier I added it because the "not whitelisted: skip "/videos/" request" implied to me that my request was blocked, that, however, was a wrong interpretation on my end.)

aduth replied:
>[Sync-handler](https://github.com/Automattic/wp-calypso/tree/master/client/lib/wp/sync-handler) is used for calypso offline. _[...]_

Thanks for the info!

aduth commented on client/state/action-types.js:
```diff
@@ -61,4 +61,8 @@ export const SUPPORT_USER_TOKEN_FETCH = 'SUPPORT_USER_TOKEN_FETCH';
 export const SUPPORT_USER_TOKEN_SET = 'SUPPORT_USER_TOKEN_SET';
 export const SUPPORT_USER_RESTORE = 'SUPPORT_USER_RESTORE';
 export const USER_RECEIVE = 'USER_RECEIVE';
+export const VIDEO_FETCH = 'VIDEO_FETCH';
```
Already some inconsistency with the action type naming in this file :frowning: See also: `_REQUEST`, `_REQUEST_SUCCESS`, `_REQUEST_FAILURE`. I'm partial to the latter, but we should really settle on a consistent approach. Thoughts?

For reference, the idea of using "request" is such that not all network requests are expecting data (even `GET` requests), so "fetch" is not always suitable.

mjuhasz replied:
In the [action-types.js](https://github.com/Automattic/wp-calypso/blob/master/client/state/action-types.js) there are examples for both `FETCH` and `REQUEST`. My thinking was that I am *fetching* existing data (a resource, like e.g. settings) but I am *requesting* an operation (like those, arguably not very RESTful actions that introduce a state change for a resource, like activate, star, or publish).
I do not feel the need to have this distinction in the language, and for me `request` covers both scenarios. I only used `fetch` because the existing code seemed to have this distinction. I am happy to agree on the term `request` and would prefer using that in my action names.

aduth replied:
>I do not feel the need to have this distinction in the language, and for me `REQUEST` covers both scenarios _[...]_ I am happy to agree on the term `REQUEST` and would prefer using that in my action names.

Cool, I'd tend to agree. Unrelated to this pull request, we'll want to see to adopting this consistency with the other action types, and investigating how well this will fit with non-retrieval requests (e.g. `POST_UPDATE_REQUEST` for saving post?)

aduth commented on client/state/videos/actions.js:
```diff
@@ -0,0 +1,49 @@
+		} );
+
+		return new Promise( ( resolve ) => {
+			wpcom.undocumented().videos( guid, ( error, data ) => {
```
If you change your undocumented `wpcom` to return the result of `this.wpcom.req.get`, you can return this call as a promise, avoiding the need for a wrapper at all.

mjuhasz replied:
Cleaner, indeed. Fixed.

aduth commented:
A few additional remarks on the general approach of this pull request:

- There's been quite a bit of back and forth on the best approach for achieving accurate dimensions for rendering videos, whereas my understanding is that we could have very quickly implemented an approach using an arbitrary width and height. This would have had an immediate improvement on the existing user experience and allowed us to iterate with improvements in subsequent pull requests. Quick wins are generally preferable to additional weeks refining the perfect implementation.
- It's my fault for suggesting that you continue with revisions to this pull request, but in retrospect, some of the additions (e.g. Redux state, query component) could have been made in separate pull requests parallel to this one, rebasing this pull request when the others had been merged.

mjuhasz commented:
> Quick wins are generally preferable to additional weeks refining the perfect implementation.

I fully agree with that. In my understanding this pull request served as a mean to familiarize myself with the calypso codebase therefore I did more exploration than what was absolutely necessary to get something tangible. In normal iterations, this could have been

1. [must have] use hardcoded height value
2. [must have] add auto resize and calculate height with a default aspect ratio
3. [nice to have] get the real video dimensions and calculate Storage Aspect Ratio
4. [nice to have]account for width and height shortcode attributes
5. [optional] account for all shortcode attributes
6. [optional] make Display Aspect Ratio available on the rest API
7. [optional] use DAR rather than SAR in the preview plugin

Each of these would have resulted in working software, improving step-by-step.

re: separate pull requests: We can still open separate pull requests, merge those, and reference back to this one which will get closed but not merged.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,146 @@
+import omitBy from 'lodash/omitBy';
+import pick from 'lodash/pick';
+import QueryString from 'querystring';
+import tinymce from 'tinymce/tinymce';
```
Unused dependency, should be removed.

mjuhasz replied:
Fixed.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,146 @@
+	constrainVideoDimensions( shortcodeWidthAttribute, shortcodeHeightAttribute ) {
+		const defaultWidth = 640;
+		const defaultAspectRatio = 16 / 9;
+		let { width, height } = this.props.video
```
For legibility's sake, I'd specify the default values in the destructuring:

```js
let {
	width = defaultWidth,
	height = defaultWidth / defaultAspectRatio
} = this.props.video;
```

mjuhasz replied:
Fixed.

aduth replied:
>Fixed.

Appears you only moved the defaults inline, but not using destructuring assignment default values as shown in the [code example from my original comment](https://github.com/Automattic/wp-calypso/pull/2267#discussion_r53621606).

That said, if `this.props.video` can be null, falling back to an empty object may or may not be more clear. What do you think?

```js
let { width = 640, height = 640 / ( 16 / 9 ) } = this.props.video || {};
```

aduth replied:
May also want to declare the `aspectRatio` as a fixed value, since only by following the logic do you understand that it's intended to be 16:9. Being explicit makes this more obvious I'd argue.

```js
const aspectRatio = 16 / 9;
let { width = 640, height = 640 / aspectRatio } = this.props.video || {};
```

mjuhasz replied:
I like your last suggestion. In addition to that, I would extract the default width like it used to be so that the magic number becomes a constant with a descriptive name.

```js
const defaultWidth = 640;
const defaultAspectRatio = 16 / 9;
let { width = defaultWidth, height = defaultWidth / defaultAspectRatio } = this.props.video || {};
```

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,146 @@
+			width = shortcodeHeightAttribute * aspectRatio;
+			height = shortcodeHeightAttribute;
+		} else if ( shortcodeWidthAttribute && shortcodeHeightAttribute ) {
+			const definedAspectRatio = shortcodeWidthAttribute / shortcodeHeightAttribute;
```
Is the purpose of this block to keep the dimensions constrained to a 16:9 aspect ratio? If so, why do we not want to allow a user to define their own custom dimensions?

mjuhasz replied:
Good thinking. We should keep the user-defined width and height.

Even though it may result in an aspect ratio that might not be the same as the Storage Aspect Ration (SAR) the user-defined dimensions should take precedence (the user knows what they are doing, we should not be too smart here).

In some cases the video might have a different Display Aspect Ratio (DAR) than SAR, so this gives the user a chance to get the ratio right.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,146 @@
+
+		return Object.assign(
+			{},
+			pick( shortcodeAttributes, ['guid', 'w', 'h'] ),
```
`pick` creates a new object, so it's not necessary to provide an empty object as the first parameter to `Object.assign`, and potentially saves you a bit of performance cost by skipping an extra object clone.

aduth replied:
Square brackets should always have a space between the opening and closing, with the one exception being of accessing an array property via string or number literal. There's a few instances in this file which should be revised. ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#arrays-and-function-calls))

mjuhasz replied:
You are absolutely right about `pick`. I also fixed some spacing issues.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,146 @@
+			guid: shortcode.attrs.numeric[0],
+			w: videoDimensions.width,
+			h: videoDimensions.height,
+			autoplay: namedAttrs.autoplay === 'true',
```
Should we be so strict with these conditions? For example, in my testing, providing `autoplay="1"` is sufficient in causing a video to autoplay.

mjuhasz replied:
I found different ways of parsing these attributes, like [accepting only a true value](https://github.com/Automattic/jetpack/blob/b5fd4cfc5dff8f59cc60b0fa8a2c8dc0edde6739/modules/videopress/class.videopress-player.php#L302), or [also "1"](https://github.com/Automattic/jetpack/blob/b5fd4cfc5dff8f59cc60b0fa8a2c8dc0edde6739/modules/videopress/class.videopress-player.php#L580) to activate boolean type of parameters.

I also found [in the shortcode api documentation](https://codex.wordpress.org/Shortcode_API) that shortcode may use single or double quotes for attribute values, or omit them entirely if the attribute value does not contain spaces.

I found no official documentation w.r.t `0`/`1` and `true`/`false` usage for boolean shortcode attributes, but both flavors seem to used in the wild.

In the end, possibilities are (for `autoplay`, as an example): `autoplay=1`, `autoplay='1'`, `autoplay="1"`, `autoplay=true`, `autoplay='true'`, `autoplay="true"`

Shortcode parser understands single/double quoted and unquoted parameter values, so that's taken care of. I extended the logic to accept `"1"` as well.

aduth replied:
>I found no official documentation w.r.t `0`/`1` and `true`/`false` usage for boolean shortcode attributes, but both flavors seem to used in the wild.

Yeah, unfortunately this is one of those things which aren't very well standardized and left to the plugin author to choose.

Your changes look good here.

aduth replied:
>In the end, possibilities are (for `autoplay`, as an example): `autoplay=1`, `autoplay='1'`, `autoplay="1"`, `autoplay=true`, `autoplay='true'`, `autoplay="true"`

Here's another one which doesn't appear to behave correctly in the editor, but autoplays on the front-end:

`[wpvideo 11acMj3O autoplay]`

mjuhasz replied:
Great finding! I looked up the [shortcode attributes documentation](https://codex.wordpress.org/Shortcode_API#Attributes) and it is not clear to me whether this format is allowed (that is, attribute name defined as value). Nevertheless it is used this way so I mirrored the existing behavior so that the in-editor preview works the same way as the front-end.

mjuhasz replied:
I added logic to parse the `permalink` shortcode attribute. While it does nothing for me at the moment, [this issue](https://github.com/Automattic/jetpack/issues/2475) suggests that it will work later.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,146 @@
+	}
+
+	onLoad() {
+		const doc = this.props.editor.iframeElement.contentDocument;
```
Nice :+1:

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,146 @@
+	onLoad() {
+		const doc = this.props.editor.iframeElement.contentDocument;
+		const script = doc.createElement( 'script' );
+		script.src = 'https://videopress.com/videopress-iframe.js';
```
At this point, how much benefit is this script providing that we couldn't implement here with a window `resize` handler? Just out of curiosity, might make sense to leave as-is.

mjuhasz replied:
The included script takes care of finding and resizing the iframe we render, and does so by maintaining the aspect ratio. At the same time it also resizes the divs embedded in the iframe for the video, the player and whatnot, and also updates the width/height attribute of the video tag (it's quite deeply nested).

Without the resize script the iframe width is computed dynamically, but we could have to compute and update its height, but that's not enough as I outlined above: we should update dimensions of elements *within* the VideoPress iframe. I would keep the existing script so that we do not add knowledge to Calypso about how VideoPress videos are rendered: we provide an iframe, the rest is taken care if by VideoPress.

aduth replied:
>I would keep the existing script so that we do not add knowledge to Calypso about how VideoPress videos are rendered: we provide an iframe, the rest is taken care if by VideoPress.

Okay, makes sense :+1:

aduth commented on client/state/videos/test/selectors.js:
```diff
@@ -0,0 +1,66 @@
+			const video = getVideo( {
+				videos: {
+					items: {
+						kUJmAcSf: {title: 'VideoPress Demo', width: 1920, height: 1080}
```
Spaces between curly brackets.

mjuhasz replied:
Fixed.

aduth commented on client/state/videos/test/selectors.js:
```diff
@@ -0,0 +1,66 @@
+} from '../selectors';
+
+describe( 'selectors', () => {
+	describe( '#getVideo()', () => {
```
Should we add a test to set expectations for if a video object is not known?

mjuhasz replied:
Yes, we should. I just did.

aduth commented:
@mjuhasz : Hopefully my comment didn't come across too harsh. It's not always easy to balance meaningful progress with iteration, evidenced by the fact that it's one of the most common pitfalls that we all continue to encounter. Of course, it's also easier for me to comment in retrospect :smile: 

The nice thing about feature flags, as you've included here, is that we can push in-progress code, allowing us to review and iterate small chunks of progress. Iteration and small pull requests are so integral to our workflow that I'd argue it's on par with (and not conflicting with) familiarizing yourself with the codebase.

I like the split you'd identified in your comment. At this point, since this is close to being ready, I think we ought continue in this pull request unless further changes prove to be too time-consuming.

mjuhasz commented:
@aduth 

> Hopefully my comment didn't come across too harsh.

On the contrary, they were constructive and appropriate.

I agree with the rest of what you said.

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,137 @@
+		const script = doc.createElement( 'script' );
+		script.src = 'https://videopress.com/videopress-iframe.js';
+		script.type = 'text/javascript';
+		doc.getElementsByTagName( 'head' )[0].appendChild( script );
```
This question had come up for me in the context of the media modal detail view, but do you think we should be smarter about whether we insert duplicate scripts? As it is, this can add up quickly when switching between editing modes:

![image](https://cloud.githubusercontent.com/assets/1779930/13263429/8818852c-da37-11e5-862f-60ab2ea0ad11.png)

mjuhasz replied:
We could detect the presence of the videopress.js script and early return with something like:
```js
		const scripts = doc.head.getElementsByTagName('script');
		for (let script of scripts) {
			if ( script.src === src ) {
				return;
			}
		}
```

However, with this code videopress.js stops working after switching back and force between the visual/html editor tabs.
By replacing the `return` statement with `doc.head.removeChild( script );` it all works again, and videopress script tags do not pile up.

videopress.js adds a `DOMContentLoaded` event listener, that is the likely culprit. Another consequence of adding this event listener is that by removing and re-adding the script the registered event listener is left behind so those do pile up.

![image](https://cloud.githubusercontent.com/assets/1017839/13439719/11b406a0-dfef-11e5-88f5-660c6c998d6c.png)

aduth replied:
>Another consequence of adding this event listener is that by removing and re-adding the script the registered event listener is left behind so those do pile up.

And each of those event handlers is executed when switching from HTML to Visual mode? I seem to recall this being the case, but that the script itself was smart enough not to duplicate work if it had already detected that it was previously initialized.

If it is worrysome (which it very well might not be), another avenue to pursue could be to render the view within its own `iframe`, such that there would only ever be one script per `iframe`. Manipulating markup of frames, especially with concern to page lifecycle events (`DOMContentLoaded`) is sometimes easier said than done, but it is do-able. See [`<ShortcodeFrame />`](https://github.com/Automattic/wp-calypso/blob/master/client/components/shortcode/frame.jsx) and [`document.write`](https://developer.mozilla.org/en-US/docs/Web/API/Document/write) for more information if you're curious.

mjuhasz replied:
Each event handler is executed but only one does real work, the others return early.

`detect` is the `DOMContentLoaded` event handler, which is smart enough not to attach resize logic to the same VideoPress iframe twice. Relevant part of the code is the `el__vp` property setting/reading:

```js
function detect() {
        var frames = document.getElementsByTagName('iframe');
        for (var i = 0; i < frames.length; i++) {
            var el = frames[i];
            if (el.__vp) continue;
            if (isVideoPress(el)) {
                attach(el);
            }
        }
    }

    function attach(el) {
        var parent = el.parentNode;
        if (!parent) return;
        var initialWidth = el.width;
        var initialHeight = el.height;
        var resize = function resize() {
            el.width = Math.min(parent.offsetWidth, initialWidth);
            el.height = initialHeight * el.width / initialWidth;
        };
        resize();
        __vpResize(el, resize);
        el.__vp = true;
    }
```

aduth commented:
Save for the few remarks noted, this is working very well for me in the browser. Nice work :+1:

aduth commented on client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx:
```diff
@@ -0,0 +1,136 @@
+		const defaultWidth = 640;
+		const defaultAspectRatio = 16 / 9;
+		let { width = defaultWidth, height = defaultWidth / defaultAspectRatio } = this.props.video || {};
+		const aspectRatio = width / height;
```
This is redundant with `defaultAspectRatio` above, yeah?

mjuhasz replied:
This is the real aspect ratio of the video which might be different from the default (i.e. 4:3, and not 16:9). If `this.props.video` is undefined then `aspectRatio` and `defaultAspectRatio` are the same, but in general this is not true.

aduth replied:
>This is the real aspect ratio of the video which might be different from the default

Oh right, I must have had my blinders on and was focusing on the default case.

aduth commented on client/state/videos/reducer.js:
```diff
@@ -0,0 +1,58 @@
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function videoRequests(state = {}, action ) {
```
```bash
$ eslint client/state/videos/reducer.js 

/Users/andrew/Documents/Code/wp-calypso/client/state/videos/reducer.js
  43:31  warning  There must be a space inside this paren  space-in-parens

✖ 1 problem (0 errors, 1 warning)
```

mjuhasz replied:
Fixed.

aduth commented:
Argh, one last hold-up. I'm seeing multiple video requests for the same endpoint when loading a post which includes a video:

![Requests](https://cloud.githubusercontent.com/assets/1779930/13541665/b25db0f4-e22b-11e5-8a06-85f03a483b37.png)

Perhaps the view (and subsequently `<QueryPosts />`) is getting remounted?

aduth commented:
This pull request will need to be updated with the merger of #3773:

- No need to specify custom `Makefile` in `client/state`
 - Instead, require your tests directly in `client/state/test/index.js`
- No need to specify Chai middleware (already configured in root test suite) 
- Do not call `nock.restore()` in your test tear down. Instead, call `nock.cleanAll()` (the former will disable `nock` altogether for the remainder of tests)

mjuhasz commented:
The branch is rebased and test are up to the new standard.


Can you elaborate on how I could reproduce the multiple requests you saw?

If I load a draft with video with debug messages enabled (`localStorage.setItem( 'debug', '*' );`)

1. `render` is called in `WpVideoView` which in turn renders `<QueryVideo>`.
2. Right after that a `wpcom:send-request sendRequest("/videos/kUJmAcSf" )` request is sent out.
3. `WpVideoView` gets re-rendered after `wpcom-proxy-request got 200 status code for URL: "/videos/kUJmAcSf"`

No other wpcom request is sent out for `"/videos/kUJmAcSf" ` according to the console log.

aduth commented:
@mjuhasz : I should have paid attention to the "Initiator" column in the network tab:

![Initiator](https://cloud.githubusercontent.com/assets/1779930/13580451/58768f60-e470-11e5-8131-bffdc1533e53.png)

I'm guessing that these requests originate from the VideoPress script. Pretty unfortunate that the additional requests are made, but not sure we can do much about it under the current approach.

apeatling commented:
@aduth @mjuhasz This has been a long time now, and seems like a great improvement. Can we get this one in? Otherwise we should close this.

aduth commented:
@mjuhasz : Do you recall if there were any blockers to merging this? We'll want to rebase, but I'd be happy to help move this one forward, especially given the issues encountered in #3832 in using the shortcode iframe for rendering.

mjuhasz commented:
@aduth VideoPress.js issues multiple requests for the same endpoint when loading a post which includes a video. Your former assessment still holds true:

> Pretty unfortunate that the additional requests are made, but not sure we can do much about it under the current approach.

lancewillett commented:
Re-opening since it'd be great to get #1730 support for improving our video offering, especially as it's a part of a paid plan.

aduth commented:
Closed in favor of #8342
